tutorials/dataframe/df106_HiggsToFourLeptons.py

```diff
@@ -0,0 +1,194 @@
+using VecF_t = const ROOT::RVec<float>&;
+using VecI_t = const ROOT::RVec<int>&;
+bool GoodElectronsAndMuons(VecI_t type, VecF_t pt, VecF_t eta, VecF_t phi, VecF_t e, VecF_t trackd0pv, VecF_t tracksigd0pv, VecF_t z0)
+{
+    for (size_t i = 0; i < type.size(); i++) {
+        ROOT::Math::PtEtaPhiEVector p(pt[i] / 1000.0, eta[i], phi[i], e[i] / 1000.0);
+        if (type[i] == 11) {
+            if (pt[i] < 7000 || abs(eta[i]) > 2.47 || abs(trackd0pv[i] / tracksigd0pv[i]) > 5 || abs(z0[i] * sin(p.Theta())) > 0.5) return false;
+        } else {
+            if (abs(trackd0pv[i] / tracksigd0pv[i]) > 5 || abs(z0[i] * sin(p.Theta())) > 0.5) return false;
+        }
+    }
+    return true;
+}
+""")
+
+for s in samples:
+    # Select electron or muon trigger
+    df[s] = df[s].Filter("trigE || trigM")
+
+    # Select events with exactly four good leptons conserving charge and lepton numbers
+    df[s] = df[s].Define("good_lep", "(abs(lep_eta) < 2.5) && (lep_pt > 5000) && (lep_ptcone30 / lep_pt < 0.3) && (lep_etcone20 / lep_pt < 0.3)")\
+                 .Filter("Sum(good_lep) == 4")\
+                 .Filter("Sum(lep_charge[good_lep]) == 0")\
+                 .Define("goodlep_sumtypes", "Sum(lep_type[good_lep])")\
+                 .Filter("goodlep_sumtypes == 44 || goodlep_sumtypes == 52 || goodlep_sumtypes == 48")
+
+    # Apply additional cuts depending on lepton flavour
+    df[s] = df[s].Filter("GoodElectronsAndMuons(lep_type[good_lep], lep_pt[good_lep], lep_eta[good_lep], lep_phi[good_lep], lep_E[good_lep], lep_trackd0pvunbiased[good_lep], lep_tracksigd0pvunbiased[good_lep], lep_z0[good_lep])")
+
+    # Create new columns with the kinematics of good leptons
+    df[s] = df[s].Define("goodlep_pt", "lep_pt[good_lep]")\
+                 .Define("goodlep_eta", "lep_eta[good_lep]")\
+                 .Define("goodlep_phi", "lep_phi[good_lep]")\
+                 .Define("goodlep_E", "lep_E[good_lep]")
+
+    # Select leptons with high transverse momentum
+    df[s] = df[s].Filter("(goodlep_pt[0] > 25000) && (goodlep_pt[1] > 15000) && (goodlep_pt[2] > 10000)")
+
+# Apply luminosity, scale factors and MC weights for simulated events
+lumi = 10064.0
+for s in samples:
+    if "data" in s:
+        df[s] = df[s].Define("weight", "1.0")
+    else:
+        df[s] = df[s].Define("weight", "scaleFactor_ELE * scaleFactor_MUON * scaleFactor_LepTRIGGER * scaleFactor_PILEUP * mcWeight * {} / {} * {}".format(xsecs[s], sumws[s], lumi))
+
+# Compute invariant mass of the four lepton system and make a histogram
+ROOT.gInterpreter.Declare("""
+using Vec_t = const ROOT::RVec<float>&;
+float ComputeInvariantMass(Vec_t pt, Vec_t eta, Vec_t phi, Vec_t e)
+{
+    ROOT::Math::PtEtaPhiEVector p1(pt[0], eta[0], phi[0], e[0]);
+    ROOT::Math::PtEtaPhiEVector p2(pt[1], eta[1], phi[1], e[1]);
+    ROOT::Math::PtEtaPhiEVector p3(pt[2], eta[2], phi[2], e[2]);
+    ROOT::Math::PtEtaPhiEVector p4(pt[3], eta[3], phi[3], e[3]);
+    return (p1 + p2 + p3 + p4).M() / 1000;
+}
```

Comment:
didn't we have helper functions for invariant masses?

Reply:
yep, but not with the energy in the back.

Reply:
only with the mass, it's a different coordinate system.

Reply:
should we have this?

Reply:
hm hard to tell. then we could also put a helper for px, py, pz and E? where to stop? and we are running out of signatures to overload properly, meaning we need a template argument or something else to find the right impl.

Reply:
I think it does not hurt to add stuff in RDFHelpers.hxx -- anything that is not data-model-specific and that users might reinvent, possibly worse, in the majority of analyses